Hyperspace Create Index Fails for csv if data is materialized before

### Describe the issue
Probable root cause of https://github.com/dotnet/spark/issues/580
Hyperspace index with dataset (CSV) and fails during the creation if data is materialized just before create index
<!--
A clear and concise description of what the issue is.
-->

### To Reproduce
<!--
Steps to reproduce the behavior:
1. Do this first '...'
2. Write this code '....'
3. Execute the code this way '....'
4. See this behavior
5. If applicable, add screenshots to help explain your problem.
-->
I could repro the issue only if I call df.show() before hs.createIndex(). If I don't call df.show(), it works fine and creates index without issue.
```
object Test extends App {
  val spark: SparkSession = ...
  
  val df = spark.read.csv("csv/*/*")
  df.show()                            // commenting this line eliminates the issue
  val hs = new Hyperspace(spark)
  hs.createIndex(df, new IndexConfig("ic2", Seq("_c0"), Seq()))
}
```
### Expected behavior
<!--
A clear and concise description of what you expected to happen.
-->
New index created and app will show in the log.
### Additional Context
<!--
Please complete the following information if applicable:
 - OS: [e.g. Windows, Linux, iOS]
 - IDE (and version) [e.g. IntelliJ v10.1, Eclipse v2.3]
 - Apache Spark Version [e.g., Apache Spark 2.4.2]
 - Platform [e.g. Local execution, Azure Synapse Analytics, Azure/AWS Databricks, HDInsight Spark]
-->

```
com.esotericsoftware.kryo.KryoException: java.util.ConcurrentModificationException
Serialization trace:
classes (sun.misc.Launcher$AppClassLoader)
classloader (java.security.ProtectionDomain)
context (java.security.AccessControlContext)
acc (org.apache.spark.util.MutableURLClassLoader)
classLoader (org.apache.hadoop.conf.Configuration)
conf (org.apache.hadoop.io.compress.BZip2Codec)
codecs (org.apache.hadoop.io.compress.CompressionCodecFactory)
codecFactory (org.apache.spark.sql.execution.datasources.csv.CSVFileFormat)
fileFormat (com.microsoft.hyperspace.index.serde.package$HadoopFsRelationWrapper)
relation (com.microsoft.hyperspace.index.serde.package$LogicalRelationWrapper)
at com.esotericsoftware.kryo.serializers.ObjectField.write(ObjectField.java:101)
at com.esotericsoftware.kryo.serializers.FieldSerializer.write(FieldSerializer.java:508)
at com.esotericsoftware.kryo.Kryo.writeObject(Kryo.java:575)
at com.esotericsoftware.kryo.serializers.ObjectField.write(ObjectField.java:79)
...
Caused by: java.util.ConcurrentModificationException
at java.util.Vector$Itr.checkForComodification(Unknown Source)
at java.util.Vector$Itr.next(Unknown Source)
at com.esotericsoftware.kryo.serializers.CollectionSerializer.write(CollectionSerializer.java:99)
at com.esotericsoftware.kryo.serializers.CollectionSerializer.write(CollectionSerializer.java:40)
at com.esotericsoftware.kryo.Kryo.writeObject(Kryo.java:575)
at com.esotericsoftware.kryo.serializers.ObjectField.write(ObjectField.java:79)
```

--- a/src/main/scala/com/microsoft/hyperspace/index/serde/LogicalPlanSerDeUtils.scala
+++ b/src/main/scala/com/microsoft/hyperspace/index/serde/LogicalPlanSerDeUtils.scala
@@ -26,6 +26,7 @@ import org.apache.spark.sql.catalyst.plans.logical._
 import org.apache.spark.sql.execution.datasources._
 import org.apache.spark.sql.execution.datasources.csv.CSVFileFormat
 import org.apache.spark.sql.execution.datasources.json.JsonFileFormat
+import org.apache.spark.sql.execution.datasources.orc.OrcFileFormat
 import org.apache.spark.sql.execution.datasources.parquet.ParquetFileFormat
 
 import com.microsoft.hyperspace.HyperspaceException
@@ -225,7 +226,7 @@ object LogicalPlanSerDeUtils {
     def unapply(fileFormat: FileFormat): Option[FileFormat] = fileFormat match {
       case CSVFileFormatWrapper => Some(new CSVFileFormat)
       case JsonFileFormatWrapper => Some(new JsonFileFormat)
-      case f: ParquetFileFormat => Some(f)
+      case _: ParquetFileFormat | _: OrcFileFormat => Some(fileFormat)
       case _ => throw HyperspaceException("Unsupported File Format found.")
     }
   }
@@ -234,7 +235,7 @@ object LogicalPlanSerDeUtils {
     def unapply(fileFormat: FileFormat): Option[FileFormat] = fileFormat match {
       case _: CSVFileFormat => Some(CSVFileFormatWrapper)
       case _: JsonFileFormat => Some(JsonFileFormatWrapper)
-      case f: ParquetFileFormat => Some(f)
+      case _: ParquetFileFormat | _: OrcFileFormat => Some(fileFormat)
       case _ => throw HyperspaceException("Unsupported File Format found.")
     }
   }

--- a/src/main/scala/com/microsoft/hyperspace/index/serde/LogicalPlanSerDeUtils.scala
+++ b/src/main/scala/com/microsoft/hyperspace/index/serde/LogicalPlanSerDeUtils.scala
@@ -24,6 +24,8 @@ import org.apache.spark.sql.SparkSession
 import org.apache.spark.sql.catalyst.expressions.{Exists, InSubquery, ListQuery, ScalarSubquery, ScalaUDF}
 import org.apache.spark.sql.catalyst.plans.logical._
 import org.apache.spark.sql.execution.datasources._
+import org.apache.spark.sql.execution.datasources.csv.CSVFileFormat
+import org.apache.spark.sql.execution.datasources.parquet.ParquetFileFormat
 
 /**
  * Utility functions for logical plan serialization/deserialization.
@@ -119,7 +121,7 @@ object LogicalPlanSerDeUtils {
             partitionSchema,
             dataSchema,
             bucketSpec,
-            fileFormat,
+            fileFormat: ParquetFileFormat,
             options),
           output,
           catalogTable,
@@ -131,6 +133,28 @@ object LogicalPlanSerDeUtils {
             dataSchema,
             bucketSpec,
             fileFormat,
+            options),
+          output,
+          catalogTable,
+          isStreaming)
+      case LogicalRelation(
+          HadoopFsRelation(
+            location: InMemoryFileIndex,
+            partitionSchema,
+            dataSchema,
+            bucketSpec,
+            _: CSVFileFormat,
+            options),
+          output,
+          catalogTable,
+          isStreaming) =>
+        LogicalRelationWrapper(
+          HadoopFsRelationWrapper(
+            InMemoryFileIndexWrapper(location.rootPaths.map(path => path.toString)),
+            partitionSchema,
+            dataSchema,
+            bucketSpec,
+            CSVFileFormatWrapper,
             options),
           output,
           catalogTable,
@@ -165,7 +189,7 @@ object LogicalPlanSerDeUtils {
             partitionSchema,
             dataSchema,
             bucketSpec,
-            fileFormat,
+            fileFormat: ParquetFileFormat,
             options),
           output,
           catalogTable,
@@ -181,6 +205,32 @@ object LogicalPlanSerDeUtils {
             dataSchema,
             bucketSpec,
             fileFormat,
+            options)(spark),
+          output,
+          catalogTable,
+          isStreaming)
+      case LogicalRelationWrapper(
+          HadoopFsRelationWrapper(
+            location: InMemoryFileIndexWrapper,
+            partitionSchema,
+            dataSchema,
+            bucketSpec,
+            CSVFileFormatWrapper,
+            options),
+          output,
+          catalogTable,
+          isStreaming) =>
+        LogicalRelation(
+          HadoopFsRelation(
+            new InMemoryFileIndex(
+              spark,
+              location.rootPathStrings.map(path => new Path(path)),
+              Map(),
+              None),
+            partitionSchema,
+            dataSchema,
+            bucketSpec,
+            new CSVFileFormat,
             options)(spark),
           output,
           catalogTable,

--- a/src/main/scala/com/microsoft/hyperspace/index/serde/LogicalPlanSerDeUtils.scala
+++ b/src/main/scala/com/microsoft/hyperspace/index/serde/LogicalPlanSerDeUtils.scala
@@ -227,7 +227,8 @@ object LogicalPlanSerDeUtils {
       case CSVFileFormatWrapper => Some(new CSVFileFormat)
       case JsonFileFormatWrapper => Some(new JsonFileFormat)
       case _: ParquetFileFormat | _: OrcFileFormat => Some(fileFormat)
-      case f => throw HyperspaceException(s"Unsupported File Format found: ${f.toString}.")
+      case other =>
+        throw HyperspaceException(s"Unsupported file format found: ${other.toString}.")
     }
   }
 
@@ -236,7 +237,8 @@ object LogicalPlanSerDeUtils {
       case _: CSVFileFormat => Some(CSVFileFormatWrapper)
       case _: JsonFileFormat => Some(JsonFileFormatWrapper)
       case _: ParquetFileFormat | _: OrcFileFormat => Some(fileFormat)
-      case f => throw HyperspaceException(s"Unsupported File Format found: ${f.toString}.")
+      case other =>
+        throw HyperspaceException(s"Unsupported file format found: ${other.toString}.")
     }
   }
 }

--- a/src/main/scala/com/microsoft/hyperspace/index/serde/LogicalPlanSerDeUtils.scala
+++ b/src/main/scala/com/microsoft/hyperspace/index/serde/LogicalPlanSerDeUtils.scala
@@ -227,7 +227,7 @@ object LogicalPlanSerDeUtils {
       case CSVFileFormatWrapper => Some(new CSVFileFormat)
       case JsonFileFormatWrapper => Some(new JsonFileFormat)
       case _: ParquetFileFormat | _: OrcFileFormat => Some(fileFormat)
-      case _ => throw HyperspaceException("Unsupported File Format found.")
+      case f => throw HyperspaceException(s"Unsupported File Format found: ${f.toString}.")
     }
   }
 
@@ -236,7 +236,7 @@ object LogicalPlanSerDeUtils {
       case _: CSVFileFormat => Some(CSVFileFormatWrapper)
       case _: JsonFileFormat => Some(JsonFileFormatWrapper)
       case _: ParquetFileFormat | _: OrcFileFormat => Some(fileFormat)
-      case _ => throw HyperspaceException("Unsupported File Format found.")
+      case f => throw HyperspaceException(s"Unsupported File Format found: ${f.toString}.")
     }
   }
 }

--- a/src/main/scala/com/microsoft/hyperspace/index/serde/LogicalPlanSerDeUtils.scala
+++ b/src/main/scala/com/microsoft/hyperspace/index/serde/LogicalPlanSerDeUtils.scala
@@ -25,6 +25,7 @@ import org.apache.spark.sql.catalyst.expressions.{Exists, InSubquery, ListQuery,
 import org.apache.spark.sql.catalyst.plans.logical._
 import org.apache.spark.sql.execution.datasources._
 import org.apache.spark.sql.execution.datasources.csv.CSVFileFormat
+import org.apache.spark.sql.execution.datasources.json.JsonFileFormat
 import org.apache.spark.sql.execution.datasources.parquet.ParquetFileFormat
 
 import com.microsoft.hyperspace.HyperspaceException
@@ -222,16 +223,18 @@ object LogicalPlanSerDeUtils {
 
   object ExtractFileFormat {
     def unapply(fileFormat: FileFormat): Option[FileFormat] = fileFormat match {
+      case CSVFileFormatWrapper => Some(new CSVFileFormat)
+      case JsonFileFormatWrapper => Some(new JsonFileFormat)
       case f: ParquetFileFormat => Some(f)
-      case _: CSVFileFormat => Some(CSVFileFormatWrapper)
       case _ => throw HyperspaceException("Unsupported File Format found.")
     }
   }
 
   object ExtractSerializableFileFormat {
     def unapply(fileFormat: FileFormat): Option[FileFormat] = fileFormat match {
+      case _: CSVFileFormat => Some(CSVFileFormatWrapper)
+      case _: JsonFileFormat => Some(JsonFileFormatWrapper)
       case f: ParquetFileFormat => Some(f)
-      case CSVFileFormatWrapper => Some(new CSVFileFormat)
       case _ => throw HyperspaceException("Unsupported File Format found.")
     }
   }

--- a/src/main/scala/com/microsoft/hyperspace/index/serde/LogicalPlanSerDeUtils.scala
+++ b/src/main/scala/com/microsoft/hyperspace/index/serde/LogicalPlanSerDeUtils.scala
@@ -123,23 +123,18 @@ object LogicalPlanSerDeUtils {
             partitionSchema,
             dataSchema,
             bucketSpec,
-            fileFormat,
+            ExtractSerializableFileFormat(fileFormat),
             options),
           output,
           catalogTable,
           isStreaming) =>
-        val format = fileFormat match {
-          case f: ParquetFileFormat => f
-          case _: CSVFileFormat => CSVFileFormatWrapper
-          case _ => throw HyperspaceException("Unsupported File Format found.")
-        }
         LogicalRelationWrapper(
           HadoopFsRelationWrapper(
             InMemoryFileIndexWrapper(location.rootPaths.map(path => path.toString)),
             partitionSchema,
             dataSchema,
             bucketSpec,
-            format,
+            fileFormat,
             options),
           output,
           catalogTable,
@@ -174,16 +169,11 @@ object LogicalPlanSerDeUtils {
             partitionSchema,
             dataSchema,
             bucketSpec,
-            fileFormat,
+            ExtractFileFormat(fileFormat),
             options),
           output,
           catalogTable,
           isStreaming) =>
-        val format = fileFormat match {
-          case f: ParquetFileFormat => f
-          case CSVFileFormatWrapper => new CSVFileFormat
-          case _ => throw HyperspaceException("Unsupported File Format found.")
-        }
         LogicalRelation(
           HadoopFsRelation(
             new InMemoryFileIndex(
@@ -227,6 +217,22 @@ object LogicalPlanSerDeUtils {
             e.query.children,
             e.query.exprId,
             e.query.childOutputs))
+    }
+  }
+
+  object ExtractFileFormat {
+    def unapply(fileFormat: FileFormat): Option[FileFormat] = fileFormat match {
+      case f: ParquetFileFormat => Some(f)
+      case _: CSVFileFormat => Some(CSVFileFormatWrapper)
+      case _ => throw HyperspaceException("Unsupported File Format found.")
+    }
+  }
+
+  object ExtractSerializableFileFormat {
+    def unapply(fileFormat: FileFormat): Option[FileFormat] = fileFormat match {
+      case f: ParquetFileFormat => Some(f)
+      case CSVFileFormatWrapper => Some(new CSVFileFormat)
+      case _ => throw HyperspaceException("Unsupported File Format found.")
     }
   }
 }

--- a/src/main/scala/com/microsoft/hyperspace/index/serde/LogicalPlanSerDeUtils.scala
+++ b/src/main/scala/com/microsoft/hyperspace/index/serde/LogicalPlanSerDeUtils.scala
@@ -27,6 +27,8 @@ import org.apache.spark.sql.execution.datasources._
 import org.apache.spark.sql.execution.datasources.csv.CSVFileFormat
 import org.apache.spark.sql.execution.datasources.parquet.ParquetFileFormat
 
+import com.microsoft.hyperspace.HyperspaceException
+
 /**
  * Utility functions for logical plan serialization/deserialization.
  */
@@ -121,40 +123,23 @@ object LogicalPlanSerDeUtils {
             partitionSchema,
             dataSchema,
             bucketSpec,
-            fileFormat: ParquetFileFormat,
-            options),
-          output,
-          catalogTable,
-          isStreaming) =>
-        LogicalRelationWrapper(
-          HadoopFsRelationWrapper(
-            InMemoryFileIndexWrapper(location.rootPaths.map(path => path.toString)),
-            partitionSchema,
-            dataSchema,
-            bucketSpec,
             fileFormat,
             options),
           output,
           catalogTable,
-          isStreaming)
-      case LogicalRelation(
-          HadoopFsRelation(
-            location: InMemoryFileIndex,
-            partitionSchema,
-            dataSchema,
-            bucketSpec,
-            _: CSVFileFormat,
-            options),
-          output,
-          catalogTable,
           isStreaming) =>
+        val format = fileFormat match {
+          case f: ParquetFileFormat => f
+          case _: CSVFileFormat => CSVFileFormatWrapper
+          case _ => throw HyperspaceException("Unsupported File Format found.")
+        }
         LogicalRelationWrapper(
           HadoopFsRelationWrapper(
             InMemoryFileIndexWrapper(location.rootPaths.map(path => path.toString)),
             partitionSchema,
             dataSchema,
             bucketSpec,
-            CSVFileFormatWrapper,
+            format,
             options),
           output,
           catalogTable,
@@ -189,11 +174,16 @@ object LogicalPlanSerDeUtils {
             partitionSchema,
             dataSchema,
             bucketSpec,
-            fileFormat: ParquetFileFormat,
+            fileFormat,
             options),
           output,
           catalogTable,
           isStreaming) =>
+        val format = fileFormat match {
+          case f: ParquetFileFormat => f
+          case CSVFileFormatWrapper => new CSVFileFormat
+          case _ => throw HyperspaceException("Unsupported File Format found.")
+        }
         LogicalRelation(
           HadoopFsRelation(
             new InMemoryFileIndex(
@@ -205,32 +195,6 @@ object LogicalPlanSerDeUtils {
             dataSchema,
             bucketSpec,
             fileFormat,
-            options)(spark),
-          output,
-          catalogTable,
-          isStreaming)
-      case LogicalRelationWrapper(
-          HadoopFsRelationWrapper(
-            location: InMemoryFileIndexWrapper,
-            partitionSchema,
-            dataSchema,
-            bucketSpec,
-            CSVFileFormatWrapper,
-            options),
-          output,
-          catalogTable,
-          isStreaming) =>
-        LogicalRelation(
-          HadoopFsRelation(
-            new InMemoryFileIndex(
-              spark,
-              location.rootPathStrings.map(path => new Path(path)),
-              Map(),
-              None),
-            partitionSchema,
-            dataSchema,
-            bucketSpec,
-            new CSVFileFormat,
             options)(spark),
           output,
           catalogTable,

--- a/src/main/scala/com/microsoft/hyperspace/index/serde/package.scala
+++ b/src/main/scala/com/microsoft/hyperspace/index/serde/package.scala
@@ -16,13 +16,14 @@
 
 package com.microsoft.hyperspace.index
 
-import org.apache.hadoop.fs.Path
-import org.apache.spark.sql.SQLContext
+import org.apache.hadoop.fs.{FileStatus, Path}
+import org.apache.hadoop.mapreduce.Job
+import org.apache.spark.sql.{SparkSession, SQLContext}
 import org.apache.spark.sql.catalyst.catalog.{BucketSpec, CatalogTable}
 import org.apache.spark.sql.catalyst.expressions.{Attribute, AttributeReference, Expression, ExprId, Predicate, Unevaluable}
 import org.apache.spark.sql.catalyst.plans.logical.{BinaryNode, LeafNode, LogicalPlan}
 import org.apache.spark.sql.execution.FileRelation
-import org.apache.spark.sql.execution.datasources.{FileFormat, FileIndex, PartitionDirectory}
+import org.apache.spark.sql.execution.datasources.{FileFormat, FileIndex, OutputWriterFactory, PartitionDirectory}
 import org.apache.spark.sql.sources.BaseRelation
 import org.apache.spark.sql.types.{DataType, StructType}
 
@@ -165,4 +166,17 @@ package object serde {
       override val isStreaming: Boolean)
       extends LeafNode
       with LogicalPlanWrapperWithInMemoryStates
+
+  case object CSVFileFormatWrapper extends FileFormat {
+    override def inferSchema(
+        sparkSession: SparkSession,
+        options: Map[String, String],
+        files: Seq[FileStatus]): Option[StructType] = throw new UnsupportedOperationException()
+
+    override def prepareWrite(
+        sparkSession: SparkSession,
+        job: Job,
+        options: Map[String, String],
+        dataSchema: StructType): OutputWriterFactory = throw new UnsupportedOperationException()
+  }
 }

--- a/src/main/scala/com/microsoft/hyperspace/index/serde/package.scala
+++ b/src/main/scala/com/microsoft/hyperspace/index/serde/package.scala
@@ -179,4 +179,17 @@ package object serde {
         options: Map[String, String],
         dataSchema: StructType): OutputWriterFactory = throw new UnsupportedOperationException()
   }
+
+  case object JsonFileFormatWrapper extends FileFormat {
+    override def inferSchema(
+      sparkSession: SparkSession,
+      options: Map[String, String],
+      files: Seq[FileStatus]): Option[StructType] = throw new UnsupportedOperationException()
+
+    override def prepareWrite(
+      sparkSession: SparkSession,
+      job: Job,
+      options: Map[String, String],
+      dataSchema: StructType): OutputWriterFactory = throw new UnsupportedOperationException()
+  }
 }

--- a/src/main/scala/com/microsoft/hyperspace/index/serde/package.scala
+++ b/src/main/scala/com/microsoft/hyperspace/index/serde/package.scala
@@ -167,20 +167,7 @@ package object serde {
       extends LeafNode
       with LogicalPlanWrapperWithInMemoryStates
 
-  case object CSVFileFormatWrapper extends FileFormat {
-    override def inferSchema(
-        sparkSession: SparkSession,
-        options: Map[String, String],
-        files: Seq[FileStatus]): Option[StructType] = throw new UnsupportedOperationException()
-
-    override def prepareWrite(
-        sparkSession: SparkSession,
-        job: Job,
-        options: Map[String, String],
-        dataSchema: StructType): OutputWriterFactory = throw new UnsupportedOperationException()
-  }
-
-  case object JsonFileFormatWrapper extends FileFormat {
+  abstract class FileFormatWrapper extends FileFormat {
     override def inferSchema(
       sparkSession: SparkSession,
       options: Map[String, String],
@@ -192,4 +179,8 @@ package object serde {
       options: Map[String, String],
       dataSchema: StructType): OutputWriterFactory = throw new UnsupportedOperationException()
   }
+
+  case object CSVFileFormatWrapper extends FileFormatWrapper
+
+  case object JsonFileFormatWrapper extends FileFormatWrapper
 }

--- a/src/test/scala/com/microsoft/hyperspace/index/LogicalPlanSerDeTests.scala
+++ b/src/test/scala/com/microsoft/hyperspace/index/LogicalPlanSerDeTests.scala
@@ -90,7 +90,7 @@ class LogicalPlanSerDeTests extends SparkFunSuite with SparkInvolvedSuite {
     val kryoSerializer = new KryoSerializer(spark.sparkContext.getConf)
     KryoSerDeUtils.serialize(kryoSerializer, fileFormat)
 
-    // Confirm that isSplittable makes serialization fail
+    // Confirm that isSplittable makes serialization fail.
     fileFormat.isSplitable(spark, Map(), new Path("path"))
     if (expectSerializationError) {
       intercept[KryoException](KryoSerDeUtils.serialize(kryoSerializer, fileFormat))
@@ -98,32 +98,28 @@ class LogicalPlanSerDeTests extends SparkFunSuite with SparkInvolvedSuite {
       KryoSerDeUtils.serialize(kryoSerializer, fileFormat)
     }
 
-    // Now verify if Hyperspace serialization still works with this format
+    // Now verify if Hyperspace serialization still works with this format.
     verifyPlanSerde(updatedScanNode, "hadoopFsRelation.plan")
   }
 
   test("Serde query with Hadoop file system parquet relation.") {
-    val parquetFormat = new ParquetFileFormat
-    // CSVFormat is unserializable due to internal unserializable members. Verify as below.
-    verifyFileFormat(parquetFormat, false)
+    // Parquet is serializable due to no internal unserializable members. Verify as below.
+    verifyFileFormat(new ParquetFileFormat, false)
   }
 
   test("Serde query with Hadoop file system csv relation.") {
-    val csvFormat = new CSVFileFormat
     // CSVFormat is unserializable due to internal unserializable members. Verify as below.
-    verifyFileFormat(csvFormat, true)
+    verifyFileFormat(new CSVFileFormat, true)
   }
 
   test("Serde query with Hadoop file system json relation.") {
-    val jsonFormat = new JsonFileFormat
     // JsonFileFormat is unserializable due to internal unserializable members. Verify as below.
-    verifyFileFormat(jsonFormat, true)
+    verifyFileFormat(new JsonFileFormat, true)
   }
 
   test("Serde query with Hadoop file system orc relation.") {
-    val orcFormat = new OrcFileFormat
-    // OrcFileFormat is serializable due to no internal unserializable members. Verify as below.
-    verifyFileFormat(orcFormat, false)
+    // Orc is serializable due to no internal unserializable members. Verify as below.
+    verifyFileFormat(new OrcFileFormat, false)
   }
 
   test("Serde query with scalar subquery.") {

--- a/src/test/scala/com/microsoft/hyperspace/index/LogicalPlanSerDeTests.scala
+++ b/src/test/scala/com/microsoft/hyperspace/index/LogicalPlanSerDeTests.scala
@@ -74,11 +74,13 @@ class LogicalPlanSerDeTests extends SparkFunSuite with SparkInvolvedSuite {
     StructType(attributes.map(a => StructField(a.name, a.dataType, a.nullable, a.metadata)))
   }
 
-  test("Serde query with Hadoop file system parquet relation.") {
+  test("Serde query with Hadoop file system relation.") {
     verifyPlanSerde(scanNode, "hadoopFsRelation.plan")
   }
 
-  private def verifyUnSerializableFileFormat(fileFormat: FileFormat): Unit = {
+  private def verifyFileFormat(
+      fileFormat: FileFormat,
+      expectSerializationError: Boolean): Unit = {
     val relation: HadoopFsRelation =
       scanNode.relation.asInstanceOf[HadoopFsRelation].copy(fileFormat = fileFormat)(spark)
     val updatedScanNode = scanNode.copy(relation = relation)
@@ -89,8 +91,10 @@ class LogicalPlanSerDeTests extends SparkFunSuite with SparkInvolvedSuite {
     KryoSerDeUtils.serialize(kryoSerializer, fileFormat)
 
     // Confirm that isSplittable makes serialization fail
-    intercept[KryoException] {
-      fileFormat.isSplitable(spark, Map(), new Path("path"))
+    fileFormat.isSplitable(spark, Map(), new Path("path"))
+    if (expectSerializationError) {
+      intercept[KryoException](KryoSerDeUtils.serialize(kryoSerializer, fileFormat))
+    } else {
       KryoSerDeUtils.serialize(kryoSerializer, fileFormat)
     }
 
@@ -98,33 +102,28 @@ class LogicalPlanSerDeTests extends SparkFunSuite with SparkInvolvedSuite {
     verifyPlanSerde(updatedScanNode, "hadoopFsRelation.plan")
   }
 
+  test("Serde query with Hadoop file system parquet relation.") {
+    val parquetFormat = new ParquetFileFormat
+    // CSVFormat is unserializable due to internal unserializable members. Verify as below.
+    verifyFileFormat(parquetFormat, false)
+  }
+
   test("Serde query with Hadoop file system csv relation.") {
     val csvFormat = new CSVFileFormat
     // CSVFormat is unserializable due to internal unserializable members. Verify as below.
-    verifyUnSerializableFileFormat(csvFormat)
+    verifyFileFormat(csvFormat, true)
   }
 
   test("Serde query with Hadoop file system json relation.") {
     val jsonFormat = new JsonFileFormat
     // JsonFileFormat is unserializable due to internal unserializable members. Verify as below.
-    verifyUnSerializableFileFormat(jsonFormat)
+    verifyFileFormat(jsonFormat, true)
   }
 
   test("Serde query with Hadoop file system orc relation.") {
     val orcFormat = new OrcFileFormat
-    val relation: HadoopFsRelation =
-      scanNode.relation.asInstanceOf[HadoopFsRelation].copy(fileFormat = orcFormat)(spark)
-    val orcScanNode = scanNode.copy(relation = relation)
-
-    // Orc file format is serializable as it doesn't contain unserializable objects like in csv or
-    // json.
-    val kryoSerializer = new KryoSerializer(spark.sparkContext.getConf)
-    KryoSerDeUtils.serialize(kryoSerializer, orcFormat)
-    orcFormat.isSplitable(spark, Map(), new Path("path"))
-    KryoSerDeUtils.serialize(kryoSerializer, orcFormat)
-
-    // Now verify if Hyperspace serialization works with orc format
-    verifyPlanSerde(orcScanNode, "hadoopFsRelation.plan")
+    // OrcFileFormat is serializable due to no internal unserializable members. Verify as below.
+    verifyFileFormat(orcFormat, false)
   }
 
   test("Serde query with scalar subquery.") {

--- a/src/test/scala/com/microsoft/hyperspace/index/LogicalPlanSerDeTests.scala
+++ b/src/test/scala/com/microsoft/hyperspace/index/LogicalPlanSerDeTests.scala
@@ -16,17 +16,20 @@
 
 package com.microsoft.hyperspace.index
 
+import com.esotericsoftware.kryo.KryoException
 import org.apache.hadoop.fs.Path
 import org.apache.spark.SparkFunSuite
+import org.apache.spark.serializer.KryoSerializer
 import org.apache.spark.sql.catalyst.expressions.{And, Attribute, AttributeReference, EqualTo, Exists, InSubquery, IsNotNull, ListQuery, Literal, NamedExpression, ScalarSubquery, ScalaUDF}
 import org.apache.spark.sql.catalyst.plans.JoinType
 import org.apache.spark.sql.catalyst.plans.logical._
 import org.apache.spark.sql.execution.datasources._
+import org.apache.spark.sql.execution.datasources.csv.CSVFileFormat
 import org.apache.spark.sql.execution.datasources.parquet.ParquetFileFormat
 import org.apache.spark.sql.types.{IntegerType, StringType, StructField, StructType}
 
 import com.microsoft.hyperspace.SparkInvolvedSuite
-import com.microsoft.hyperspace.index.serde.LogicalPlanSerDeUtils
+import com.microsoft.hyperspace.index.serde.{KryoSerDeUtils, LogicalPlanSerDeUtils}
 
 /**
  * Some tests are adapted from examples in ExpressionParserSuite.scala, PlanParserSuite.scala,
@@ -40,6 +43,7 @@ class LogicalPlanSerDeTests extends SparkFunSuite with SparkInvolvedSuite {
 
   var scanNode: LogicalRelation = _
   var singleTablePlan: LogicalPlan = _
+  var csvScanNode: LogicalRelation = _
 
   override def beforeAll(): Unit = {
     super.beforeAll()
@@ -63,14 +67,41 @@ class LogicalPlanSerDeTests extends SparkFunSuite with SparkInvolvedSuite {
     singleTablePlan = Project(
       Seq(c1, c2, c3),
       Filter(And(IsNotNull(c3), EqualTo(c3, Literal("facebook"))), scanNode))
+
+    val csvRelation = HadoopFsRelation(
+      tableLocation,
+      new StructType(),
+      tableSchema,
+      None,
+      new CSVFileFormat,
+      Map.empty)(spark)
+    val csvScanNode = LogicalRelation(csvRelation, Seq(c1, c2, c3, c4), None, isStreaming = false)
   }
 
   private def schemaFromAttributes(attributes: Attribute*): StructType = {
     StructType(attributes.map(a => StructField(a.name, a.dataType, a.nullable, a.metadata)))
   }
 
-  test("Serde query with Hadoop file system relation.") {
+  test("Serde query with Hadoop file system parquet relation.") {
     verifyPlanSerde(scanNode, "hadoopFsRelation.plan")
+  }
+
+  test("Serde query with Hadoop file system csv relation.") {
+    val csvFormat = csvScanNode.relation.asInstanceOf[HadoopFsRelation].fileFormat
+
+    // Csv file format is serializable unless isSplittable is called on it. isSplittable api
+    // initializes internal objects which break serialization logic.
+    val kryoSerializer = new KryoSerializer(spark.sparkContext.getConf)
+    KryoSerDeUtils.serialize(kryoSerializer, csvFormat)
+
+    // Confirm that isSplittable makes serialization fail
+    intercept[KryoException] {
+      csvFormat.isSplitable(spark, Map(), new Path("path"))
+      KryoSerDeUtils.serialize(kryoSerializer, csvFormat)
+    }
+
+    // Now verify if Hyperspace serialization still works with csv format
+    verifyPlanSerde(csvScanNode, "hadoopFsRelation.plan")
   }
 
   test("Serde query with scalar subquery.") {

--- a/src/test/scala/com/microsoft/hyperspace/index/LogicalPlanSerDeTests.scala
+++ b/src/test/scala/com/microsoft/hyperspace/index/LogicalPlanSerDeTests.scala
@@ -43,7 +43,6 @@ class LogicalPlanSerDeTests extends SparkFunSuite with SparkInvolvedSuite {
 
   var scanNode: LogicalRelation = _
   var singleTablePlan: LogicalPlan = _
-  var csvScanNode: LogicalRelation = _
 
   override def beforeAll(): Unit = {
     super.beforeAll()
@@ -67,15 +66,6 @@ class LogicalPlanSerDeTests extends SparkFunSuite with SparkInvolvedSuite {
     singleTablePlan = Project(
       Seq(c1, c2, c3),
       Filter(And(IsNotNull(c3), EqualTo(c3, Literal("facebook"))), scanNode))
-
-    val csvRelation = HadoopFsRelation(
-      tableLocation,
-      new StructType(),
-      tableSchema,
-      None,
-      new CSVFileFormat,
-      Map.empty)(spark)
-    val csvScanNode = LogicalRelation(csvRelation, Seq(c1, c2, c3, c4), None, isStreaming = false)
   }
 
   private def schemaFromAttributes(attributes: Attribute*): StructType = {
@@ -87,7 +77,10 @@ class LogicalPlanSerDeTests extends SparkFunSuite with SparkInvolvedSuite {
   }
 
   test("Serde query with Hadoop file system csv relation.") {
-    val csvFormat = csvScanNode.relation.asInstanceOf[HadoopFsRelation].fileFormat
+    val csvFormat = new CSVFileFormat
+    val csvRelation: HadoopFsRelation =
+      scanNode.relation.asInstanceOf[HadoopFsRelation].copy(fileFormat = csvFormat)(spark)
+    val csvScanNode = scanNode.copy(relation = csvRelation)
 
     // Csv file format is serializable unless isSplittable is called on it. isSplittable api
     // initializes internal objects which break serialization logic.

--- a/src/test/scala/com/microsoft/hyperspace/index/LogicalPlanSerDeTests.scala
+++ b/src/test/scala/com/microsoft/hyperspace/index/LogicalPlanSerDeTests.scala
@@ -78,46 +78,36 @@ class LogicalPlanSerDeTests extends SparkFunSuite with SparkInvolvedSuite {
     verifyPlanSerde(scanNode, "hadoopFsRelation.plan")
   }
 
-  test("Serde query with Hadoop file system csv relation.") {
-    val csvFormat = new CSVFileFormat
+  private def verifyUnSerializableFileFormat(fileFormat: FileFormat): Unit = {
     val relation: HadoopFsRelation =
-      scanNode.relation.asInstanceOf[HadoopFsRelation].copy(fileFormat = csvFormat)(spark)
-    val csvScanNode = scanNode.copy(relation = relation)
+      scanNode.relation.asInstanceOf[HadoopFsRelation].copy(fileFormat = fileFormat)(spark)
+    val updatedScanNode = scanNode.copy(relation = relation)
 
-    // Csv file format is serializable unless isSplittable is called on it. isSplittable api
-    // initializes internal objects which break serialization logic.
+    // This file format is serializable unless isSplittable is called on it. isSplittable api
+    // initializes some unserializable internal objects which break the serialization logic.
     val kryoSerializer = new KryoSerializer(spark.sparkContext.getConf)
-    KryoSerDeUtils.serialize(kryoSerializer, csvFormat)
+    KryoSerDeUtils.serialize(kryoSerializer, fileFormat)
 
     // Confirm that isSplittable makes serialization fail
     intercept[KryoException] {
-      csvFormat.isSplitable(spark, Map(), new Path("path"))
-      KryoSerDeUtils.serialize(kryoSerializer, csvFormat)
+      fileFormat.isSplitable(spark, Map(), new Path("path"))
+      KryoSerDeUtils.serialize(kryoSerializer, fileFormat)
     }
 
-    // Now verify if Hyperspace serialization still works with csv format
-    verifyPlanSerde(csvScanNode, "hadoopFsRelation.plan")
+    // Now verify if Hyperspace serialization still works with this format
+    verifyPlanSerde(updatedScanNode, "hadoopFsRelation.plan")
+  }
+
+  test("Serde query with Hadoop file system csv relation.") {
+    val csvFormat = new CSVFileFormat
+    // CSVFormat is unserializable due to internal unserializable members. Verify as below.
+    verifyUnSerializableFileFormat(csvFormat)
   }
 
   test("Serde query with Hadoop file system json relation.") {
     val jsonFormat = new JsonFileFormat
-    val relation: HadoopFsRelation =
-      scanNode.relation.asInstanceOf[HadoopFsRelation].copy(fileFormat = jsonFormat)(spark)
-    val jsonScanNode = scanNode.copy(relation = relation)
-
-    // Json file format is serializable unless isSplittable is called on it. isSplittable api
-    // initializes internal objects which break serialization logic.
-    val kryoSerializer = new KryoSerializer(spark.sparkContext.getConf)
-    KryoSerDeUtils.serialize(kryoSerializer, jsonFormat)
-
-    // Confirm that isSplittable makes serialization fail
-    intercept[KryoException] {
-      jsonFormat.isSplitable(spark, Map(), new Path("path"))
-      KryoSerDeUtils.serialize(kryoSerializer, jsonFormat)
-    }
-
-    // Now verify if Hyperspace serialization still works with json format
-    verifyPlanSerde(jsonScanNode, "hadoopFsRelation.plan")
+    // JsonFileFormat is unserializable due to internal unserializable members. Verify as below.
+    verifyUnSerializableFileFormat(jsonFormat)
   }
 
   test("Serde query with Hadoop file system orc relation.") {
@@ -126,7 +116,8 @@ class LogicalPlanSerDeTests extends SparkFunSuite with SparkInvolvedSuite {
       scanNode.relation.asInstanceOf[HadoopFsRelation].copy(fileFormat = orcFormat)(spark)
     val orcScanNode = scanNode.copy(relation = relation)
 
-    // Orc file format is serializable by default, so materialization should not affect.
+    // Orc file format is serializable as it doesn't contain unserializable objects like in csv or
+    // json.
     val kryoSerializer = new KryoSerializer(spark.sparkContext.getConf)
     KryoSerDeUtils.serialize(kryoSerializer, orcFormat)
     orcFormat.isSplitable(spark, Map(), new Path("path"))

--- a/src/test/scala/com/microsoft/hyperspace/index/LogicalPlanSerDeTests.scala
+++ b/src/test/scala/com/microsoft/hyperspace/index/LogicalPlanSerDeTests.scala
@@ -26,6 +26,7 @@ import org.apache.spark.sql.catalyst.plans.logical._
 import org.apache.spark.sql.execution.datasources._
 import org.apache.spark.sql.execution.datasources.csv.CSVFileFormat
 import org.apache.spark.sql.execution.datasources.json.JsonFileFormat
+import org.apache.spark.sql.execution.datasources.orc.OrcFileFormat
 import org.apache.spark.sql.execution.datasources.parquet.ParquetFileFormat
 import org.apache.spark.sql.types.{IntegerType, StringType, StructField, StructType}
 
@@ -116,6 +117,22 @@ class LogicalPlanSerDeTests extends SparkFunSuite with SparkInvolvedSuite {
     }
 
     // Now verify if Hyperspace serialization still works with json format
+    verifyPlanSerde(csvScanNode, "hadoopFsRelation.plan")
+  }
+
+  test("Serde query with Hadoop file system orc relation.") {
+    val orcFormat = new OrcFileFormat
+    val relation: HadoopFsRelation =
+      scanNode.relation.asInstanceOf[HadoopFsRelation].copy(fileFormat = orcFormat)(spark)
+    val csvScanNode = scanNode.copy(relation = relation)
+
+    // Orc file format is serializable by default, so materialization should not affect.
+    val kryoSerializer = new KryoSerializer(spark.sparkContext.getConf)
+    KryoSerDeUtils.serialize(kryoSerializer, orcFormat)
+    orcFormat.isSplitable(spark, Map(), new Path("path"))
+    KryoSerDeUtils.serialize(kryoSerializer, orcFormat)
+
+    // Now verify if Hyperspace serialization still works with csv format
     verifyPlanSerde(csvScanNode, "hadoopFsRelation.plan")
   }
 

--- a/src/test/scala/com/microsoft/hyperspace/index/LogicalPlanSerDeTests.scala
+++ b/src/test/scala/com/microsoft/hyperspace/index/LogicalPlanSerDeTests.scala
@@ -81,7 +81,7 @@ class LogicalPlanSerDeTests extends SparkFunSuite with SparkInvolvedSuite {
   private def verifyFileFormat(
       fileFormat: FileFormat,
       expectSerializationError: Boolean): Unit = {
-    val relation: HadoopFsRelation =
+    val relation =
       scanNode.relation.asInstanceOf[HadoopFsRelation].copy(fileFormat = fileFormat)(spark)
     val updatedScanNode = scanNode.copy(relation = relation)
 

--- a/src/test/scala/com/microsoft/hyperspace/index/LogicalPlanSerDeTests.scala
+++ b/src/test/scala/com/microsoft/hyperspace/index/LogicalPlanSerDeTests.scala
@@ -103,7 +103,7 @@ class LogicalPlanSerDeTests extends SparkFunSuite with SparkInvolvedSuite {
     val jsonFormat = new JsonFileFormat
     val relation: HadoopFsRelation =
       scanNode.relation.asInstanceOf[HadoopFsRelation].copy(fileFormat = jsonFormat)(spark)
-    val csvScanNode = scanNode.copy(relation = relation)
+    val jsonScanNode = scanNode.copy(relation = relation)
 
     // Json file format is serializable unless isSplittable is called on it. isSplittable api
     // initializes internal objects which break serialization logic.
@@ -117,14 +117,14 @@ class LogicalPlanSerDeTests extends SparkFunSuite with SparkInvolvedSuite {
     }
 
     // Now verify if Hyperspace serialization still works with json format
-    verifyPlanSerde(csvScanNode, "hadoopFsRelation.plan")
+    verifyPlanSerde(jsonScanNode, "hadoopFsRelation.plan")
   }
 
   test("Serde query with Hadoop file system orc relation.") {
     val orcFormat = new OrcFileFormat
     val relation: HadoopFsRelation =
       scanNode.relation.asInstanceOf[HadoopFsRelation].copy(fileFormat = orcFormat)(spark)
-    val csvScanNode = scanNode.copy(relation = relation)
+    val orcScanNode = scanNode.copy(relation = relation)
 
     // Orc file format is serializable by default, so materialization should not affect.
     val kryoSerializer = new KryoSerializer(spark.sparkContext.getConf)
@@ -132,8 +132,8 @@ class LogicalPlanSerDeTests extends SparkFunSuite with SparkInvolvedSuite {
     orcFormat.isSplitable(spark, Map(), new Path("path"))
     KryoSerDeUtils.serialize(kryoSerializer, orcFormat)
 
-    // Now verify if Hyperspace serialization still works with csv format
-    verifyPlanSerde(csvScanNode, "hadoopFsRelation.plan")
+    // Now verify if Hyperspace serialization works with orc format
+    verifyPlanSerde(orcScanNode, "hadoopFsRelation.plan")
   }
 
   test("Serde query with scalar subquery.") {

--- a/src/test/scala/com/microsoft/hyperspace/index/LogicalPlanSerDeTests.scala
+++ b/src/test/scala/com/microsoft/hyperspace/index/LogicalPlanSerDeTests.scala
@@ -25,6 +25,7 @@ import org.apache.spark.sql.catalyst.plans.JoinType
 import org.apache.spark.sql.catalyst.plans.logical._
 import org.apache.spark.sql.execution.datasources._
 import org.apache.spark.sql.execution.datasources.csv.CSVFileFormat
+import org.apache.spark.sql.execution.datasources.json.JsonFileFormat
 import org.apache.spark.sql.execution.datasources.parquet.ParquetFileFormat
 import org.apache.spark.sql.types.{IntegerType, StringType, StructField, StructType}
 
@@ -78,9 +79,9 @@ class LogicalPlanSerDeTests extends SparkFunSuite with SparkInvolvedSuite {
 
   test("Serde query with Hadoop file system csv relation.") {
     val csvFormat = new CSVFileFormat
-    val csvRelation: HadoopFsRelation =
+    val relation: HadoopFsRelation =
       scanNode.relation.asInstanceOf[HadoopFsRelation].copy(fileFormat = csvFormat)(spark)
-    val csvScanNode = scanNode.copy(relation = csvRelation)
+    val csvScanNode = scanNode.copy(relation = relation)
 
     // Csv file format is serializable unless isSplittable is called on it. isSplittable api
     // initializes internal objects which break serialization logic.
@@ -94,6 +95,27 @@ class LogicalPlanSerDeTests extends SparkFunSuite with SparkInvolvedSuite {
     }
 
     // Now verify if Hyperspace serialization still works with csv format
+    verifyPlanSerde(csvScanNode, "hadoopFsRelation.plan")
+  }
+
+  test("Serde query with Hadoop file system json relation.") {
+    val jsonFormat = new JsonFileFormat
+    val relation: HadoopFsRelation =
+      scanNode.relation.asInstanceOf[HadoopFsRelation].copy(fileFormat = jsonFormat)(spark)
+    val csvScanNode = scanNode.copy(relation = relation)
+
+    // Json file format is serializable unless isSplittable is called on it. isSplittable api
+    // initializes internal objects which break serialization logic.
+    val kryoSerializer = new KryoSerializer(spark.sparkContext.getConf)
+    KryoSerDeUtils.serialize(kryoSerializer, jsonFormat)
+
+    // Confirm that isSplittable makes serialization fail
+    intercept[KryoException] {
+      jsonFormat.isSplitable(spark, Map(), new Path("path"))
+      KryoSerDeUtils.serialize(kryoSerializer, jsonFormat)
+    }
+
+    // Now verify if Hyperspace serialization still works with json format
     verifyPlanSerde(csvScanNode, "hadoopFsRelation.plan")
   }
 

